src/core/lib/layer.js

```diff
@@ -480,10 +487,67 @@ export default class Layer {
     return redraw;
   }
 
+  // Helper methods
+
+  // Dirty some change flags, will be handled by updateLayer
+  setChangeFlags(flags) {
+    this.state.changeFlags = this.state.changeFlags || {};
+    const changeFlags = this.state.changeFlags;
+    if (flags.dataChanged && !changeFlags.dataChanged) {
+      changeFlags.dataChanged = flags.dataChanged;
+      changeFlags.propsOrDataChanged = flags.dataChanged;
+      changeFlags.somethingChanged = flags.dataChanged;
+      log.log(LOG_PRIORITY_UPDATE,
+        `dataChanged: ${flags.dataChanged} in ${this.id}`);
+    }
+    if (flags.updateTriggersChanged && !changeFlags.dataChanged) {
+      changeFlags.updateTriggersChanged = flags.updateTriggersChanged;
+      changeFlags.propsOrDataChanged = flags.updateTriggersChanged;
+      changeFlags.somethingChanged = flags.updateTriggersChanged;
```

Comment:
This will potentially override `propsOrDataChanged` and `somethingChanged`. Depending on the sources and order that this function is called, the output is somewhat unpredictable. (e.g. sometimes `'data changed'`, sometimes `true`)
Will it be better to assign these two fields outside of the if statements from the final `changeFlags` state?

Reply:
> This will potentially override propsOrDataChanged and somethingChanged. Depending on the sources and order that this function is called, the output is somewhat unpredictable. (e.g. sometimes 'data changed', sometimes true)

My direction here (this is a set of linked PRs in the interest of keeping things in reviewable pieces) is to drop `reason` completely and always have the actual flags be strings or `false`. It simplifies the code across the board assuming all change detecting code follows the convention, and as far as I can tell there is no value gained in keeping separate track of a boolean and a string.

> Will it be better to assign these two fields outside of the if statements from the final changeFlags state?

Well the whole point of this PR is to disconnect changeFlags from updateState. `setChangeFlags` can now be called repeatedly and must be able to gradually flip on more flags until an update clears it. Since this function the only interface to modifying changeFlags makes it natural to manage them here, why burden the caller with that?

Reply:
The flags should be flipped in `setChangedFlags`, but outside of the `if` statements.
Of course, if we drop `reason`, then it doesn't matter. Though I do believe that each PR should work without assuming a future feature change.